crypto/s2n_certificate.c

```diff
@@ -167,30 +169,131 @@ struct s2n_cert_chain_and_key *s2n_cert_chain_and_key_new(void)
 
-    chain_and_key->san_names = san_names;
+    for (int i = 0; i < sk_GENERAL_NAME_num(san_names); i++) {
+        GENERAL_NAME *san_name = sk_GENERAL_NAME_value(san_names, i);
+        if (!san_name) {
+            continue;
+        }
 
+        if (san_name->type == GEN_DNS) {
+            /* Per https://www.openssl.org/docs/man1.1.0/man3/ASN1_STRING_data.html there may
+             * be embedded NULLs inside of the SAN string. The comparison later with any proper
+             * DNS hostnames will safely fail.
+             */
+            unsigned char *san_str = san_name->d.dNSName->data;
+            const size_t san_str_len = san_name->d.dNSName->length;
+            struct s2n_blob *san_blob = s2n_array_insert(chain_and_key->san_names, s2n_array_num_elements(chain_and_key->san_names));
+            if (!san_blob) {
+                GENERAL_NAMES_free(san_names);
+                return -1;
+            }
+
+            if (s2n_alloc(san_blob, san_str_len)) {
+                GENERAL_NAMES_free(san_names);
+                return -1;
+            }
+
+            memcpy_check(san_blob->data, san_str, san_str_len);
+            san_blob->size = san_str_len;
+        }
+    }
+
+    GENERAL_NAMES_free(san_names);
+    return 0;
+}
+
+static int s2n_cert_chain_and_key_load_cn(struct s2n_cert_chain_and_key *chain_and_key, X509 *x509_cert)
+{
+    notnull_check(chain_and_key->cn_names);
+
+    X509_NAME *subject = X509_get_subject_name(x509_cert);
+    if (!subject) {
+        return 0;
+    }
+
+    /* Technically there can be multiple CommonNames entries in the Subject. */
+    int lastpos = -1;
+    while((lastpos = X509_NAME_get_index_by_NID(subject, NID_commonName, lastpos)) >= 0) {
+        X509_NAME_ENTRY *name_entry = X509_NAME_get_entry(subject, lastpos);
+        if (!name_entry) {
+            continue;
+        }
+
+        ASN1_STRING *asn1_str = X509_NAME_ENTRY_get_data(name_entry);
+        if (!asn1_str) {
+            continue;
+        }
+
+        /* We need to try and decode the CN since it may be encoded as unicode with a
+         * direct ASCII equivalent. Any non ASCII bytes in the string will fail later when we
+         * actually compare hostnames.
+         */
+        unsigned char *utf8_str;
+        const int utf8_out_len = ASN1_STRING_to_UTF8(&utf8_str, asn1_str);
+        if (utf8_out_len < 0) {
+            continue;
+        }
+
+        if (utf8_out_len > 0) {
+            struct s2n_blob *cn_name = s2n_array_insert(chain_and_key->cn_names, s2n_array_num_elements(chain_and_key->cn_names));
+            if (cn_name == NULL) {
+                OPENSSL_free(utf8_str);
+                return -1;
+            }
+
+            if (s2n_alloc(cn_name, utf8_out_len) < 0) {
+                OPENSSL_free(utf8_str);
+                return -1;
+            }
+            memcpy_check(cn_name->data, utf8_str, utf8_out_len);
+            cn_name->size = utf8_out_len;
+        }
+
+        OPENSSL_free(utf8_str);
```

Comment:
note this is the case where `utf8_out_len == 0`, even though the string is 0 length, openssl does allocate memory, see https://community.openvpn.net/openvpn/wiki/VulnerabilitiesFixedInOpenVPN243 . I'm debating just doing an:

```
if (utf8_out_len < 0) {
} else if (utf8_out_len == 0) {
} else {
}
```

block here.